harness/determined/workload.py

```diff
@@ -55,8 +55,14 @@ class Skipped:
     pass
 
 
+class WorkerFailed:
```

Comment:
If that is the case, then none of `workload.py`, `_workload_manager.py`, or `_socket_manager.py` should be a part of this PR.

Why not take the following approach:
- Use a try/except in the broadcast client to detect exceptions and send the `_SerialInterruptMessage` message (maybe rename it to "ExceptionAlertMessage" or something)
- Raise an exception in `_recv_all_with_polling()` in the SubprocessLauncher as soon as you see the `_SerialInterruptMessage`
- Leave all the framework logic and all the higher level layers alone

This comes with the side effect that the ZMQBroadcastServer/Client would be even more tightly bound to dtrain communication, but I'm 100% ok with that, considering the complexity cost of not doing it that way.  And considering that that's the only we way use those classes anyway.

Reply:
I believe the approach you proposed is viable but as you said it does increase the coupling. 

The current implementation I have does not have as much coupling and I disagree with you statement that: 
> none of workload.py, _workload_manager.py, or _socket_manager.py should be a part of this PR.

`WorkerFailed` is no different than a `Skipped` message, and should be handled in similar fashion.

Reply:
I am on board with renaming `_SerialInterruptMessage` although `ExceptionAlertMessage` seems like it's overfitting the current use case a bit too much.

Reply:
> WorkerFailed is no different than a Skipped message, and should be handled in similar fashion.

A `SkippedMessage` is a sort of necessary evil since we have an architecture where non-chief machines do not respond to the master at all.  I would prefer that we didn't have it at all, but for various reasons it does exist.

The primary difference is that `SkippedMessage` is a normal message that is to be expected, and hence deserves to be part of the set of messages that can be passed through our system.  In contrast, `WorkerFailed` a message is confusing error information (which is normally handled through exceptions) with normal workflows (which are handled through message passing between layers of the harness).

Think of it this way: if we accidentally divided by zero in the WorkloadManager, we wouldn't pass a message up the stack, we'd just raise an error.  We should as much as possible strive for our inter-process error handling to mirror the single-process error handling case.  We already have that for when the health check fails; we don't pass a message we raise an error.  Why would this one particular error deserve to be a message instead of just a normal error?

Furthermore, if we rewrote the BroadcastServer/Client using TCP (which would not be that big of deal, since it's only used in one place and it's only used in one way), then we wouldn't need to introduce any new messages at all; we would just detect the connection getting broken.

But importantly, the broken connection is normally out-of-band information compared to the rest of the message stream.  And it is perfectly reasonable to communicate that out-of-band here as well; only here "out-of-band" actually means a separate message, much the same way that we originally had to simulate connections with a ConnectedMessage (although it looks like since we use the ReadMessage as a sync point we actually don't use ConnectedMessage).  You could call it "BrokenConnectionMessage" message if you'd like.

Reply:
> Furthermore, if we rewrote the BroadcastServer/Client using TCP 

I mentioned this before but changing from ZMQ to TCP isn't a discussion that belongs in a PR.

> Think of it this way: if we accidentally divided by zero in the WorkloadManager, we wouldn't pass a message up the stack, we'd just raise an error. 

This doesn't seem relevant here.

> We already have that for when the health check fails; we don't pass a message we raise an error. Why would this one particular error deserve to be a message instead of just a normal error?

For the case of a failed worker we raise the same exact error we as we do in a health check.

 > In contrast, WorkerFailed a message is confusing error information (which is normally handled through exceptions) with normal workflows (which are handled through message passing between layers of the harness).

This seems as a very arbitrary. In both the `SkippedMessage` case and the `WorkerFailed` case we are indicating that the lower level is not communicating results that are anticipated (due to different causes). They should be handled in a similar manner.